### PR TITLE
Make the VM state driver unit testable

### PR DESCRIPTION
**Pull request by gjcolombo**
Refactor the VM controller so that the worker thread that accepts incoming commands (things like 'run', 'halt', or 'migrate') has well-defined interfaces to access VM controller state, to issue commands to vCPU tasks, and to interoperate with live migration tasks. This makes it possible to write test doubles for substantially all of the state worker's side-effect-producing entanglements, which in turn makes it much more feasible to unit test the state worker.

Use this power to write some tests, including some (currently commented-out) test cases that repro issues #250, #252, and #258. This should make it much faster and easier to verify the fixes for these bugs.

It may be easier to read this PR one commit at a time. See the commit messages for each individual commit for more context about each step in the refactoring process.

### CODE REVIEW

File: bin/propolis-server/src/lib/vm/state_driver.rs [hunk 1]
```diff
@@ -0,0 +1,812 @@
...
+    }
+
+    fn reset_vcpus(&self) {
+        self.vcpu_tasks.new_generation();
+        self.controller.reset_vcpu_state();
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (luqmana) [inline bin/propolis-server/src/lib/vm/state_driver.rs:450]: woo! love these new tests! great to be able to verify the sequencing

**Review by luqmana — APPROVED** [review]
Thanks for refactoring this greg!

File: bin/propolis-server/src/lib/vm/mod.rs [hunk 2]
```diff
@@ -1,27 +1,25 @@
...
-//! and to reboot or halt an instance are handled on this worker thread.
+//! The controller wraps all of the objects describing a VM (its Propolis
+//! `Instance`, its instance spec, etc.) and a single "state driver" thread that
+//! changes the VM's state at runtime. External requests to change a VM's state
+//! (originating from a Propolis server) first reach the VM controller, which
+//! vets those requests, rejects any that are infeasible in light of the VM's
+//! current state or the pendency of prior requests, and passes the rest to the
+//! state driver for processing. As part of its work, the driver may need to
+//! request changes to the instance (e.g. pausing or resuming entities), which
+//! requests it issues by calling back into the VM controller.
```
[the remark this answers is not in the document]
Comment (gjcolombo) [reply bin/propolis-server/src/lib/vm/mod.rs]: Tidied this up a bit in f093e0c. I still have some work I want to do in here to more clearly divide responsibilities between the VM controller and the state driver, which should make it possible to clarify this theory statement even further, but I think this should make things a little clearer for now.

**Comment by gjcolombo** [comment]
Thanks for the review!

**Pull request merged**